index.test.js

```
@@ -0,0 +1,319 @@
+'use strict';
+
+const request = require('supertest');
+const Boom = require('@hapi/boom');
+const Joi = require('@hapi/joi');
+const { apiResolver } = require('next/dist/next-server/server/api-utils');
+const withRest = require('./');
+
+const enhance = (handler) => (req, res) => apiResolver(req, res, undefined, handler);
+
+beforeEach(() => {
+    console.error.mock && console.error.mockRestore();
+});
+
+describe('withRest', () => {
+    it('should respond what the handler returned', async () => {
+        const app = withRest({
+            GET: () => ({ foo: 'bar' }),
+        });
+
+        await request(enhance(app))
+        .get('/')
+        .expect(200)
+        .expect('Content-Type', /^application\/json/)
+        .then((res) => {
+            expect(res.body).toEqual({ foo: 'bar' });
+        });
+    });
+
+    it('should support nullish responses', async () => {
+        let app;
+
+        app = withRest({
+            GET: () => null,
+        });
+
+        await request(enhance(app))
+        .get('/')
+        .expect(200)
+        .expect('Content-Type', /^application\/json/)
+        .then((res) => {
+            expect(res.body).toBe(null);
+        });
+
+        app = withRest({
+            GET: () => null,
+        });
+
+        await request(enhance(app))
+        .get('/')
+        .expect(200)
+        .expect('Content-Type', /^application\/json/)
+        .then((res) => {
+            expect(res.body).toBe(null);
+        });
+    });
+
+    it('should do nothing if headers were already sent', async () => {
```

How are you validating that it should do nothing? 😕

---

I'm using `req.headersSent` to check that. This means that if you use `res.send` or `res.json` to send the respond directly in your handler, I won't do nothing.

---

I know. But it's not tested with an expect 🙃

---

You are right. I've added a return value to make it clear that it's being ignored.

---

Hmmm.. what happens in this situation? Perhaps if we return a value that isn't `undefined` in this situation, we can't really "throw" .. but perhaps we can instead throw a global error.

---

Handled by just logging an error in this case.